source/MaterialXRuntime/RtFileIo.cpp

```diff
@@ -112,36 +98,40 @@ namespace
             {
                 const RtToken portName(elem->getName());
                 const RtToken portType(elem->getType());
-                const RtValue portValue(readValue(elem->getValue(), portType, nodedef->getValueStorage()));
+                const string& valueStr = elem->getValueString();
 
-                PrvObjectHandle inputH = PrvPortDef::createNew(portName, portType, portValue);
+                PrvObjectHandle inputH = PrvPortDef::createNew(nodedef, portName, portType);
                 PrvPortDef* input = inputH->asA<PrvPortDef>();
+                if (!valueStr.empty())
+                {
+                    RtValue::fromString(portType, valueStr, input->getValue());
+                }
                 input->setColorSpace(RtToken(elem->getColorSpace()));
                 // TODO: fix when units are implemented in core
                 // input->setUnit(RtToken(elem->getUnit()));
                 readAttributes(elem, input, portdefIgnoreAttr);
-                nodedef->addPort(inputH);
             }
             else if (elem->isA<Parameter>())
             {
                 const RtToken portName(elem->getName());
                 const RtToken portType(elem->getType());
-                const RtValue portValue(readValue(elem->getValue(), portType, nodedef->getValueStorage()));
+                const string& valueStr = elem->getValueString();
 
-                PrvObjectHandle inputH = PrvPortDef::createNew(portName, portType, portValue, RtPortFlag::UNIFORM);
+                PrvObjectHandle inputH = PrvPortDef::createNew(nodedef, portName, portType, RtPortFlag::UNIFORM);
                 PrvPortDef* input = inputH->asA<PrvPortDef>();
+                if (!valueStr.empty())
+                {
+                    RtValue::fromString(portType, valueStr, input->getValue());
+                }
                 input->setColorSpace(RtToken(elem->getColorSpace()));
                 // TODO: fix when units are implemented in core
                 // input->setUnit(RtToken(elem->getUnit()));
                 readAttributes(elem, input, portdefIgnoreAttr);
-                nodedef->addPort(inputH);
             }
         }
-
-        return nodedefH;
     }
 
-    PrvObjectHandle readNode(const NodePtr& src, PrvStage* stage, PrvStage* searchStage)
+    PrvNode* readNode(const NodePtr& src, PrvElement* parent, PrvStage* stage)
     {
         NodeDefPtr srcNodedef = src->getNodeDef();
```

Comment:
This will still fail when reading in a document which does not contain this nodedef. I don't really like the extra search stage but the problem does not go away by removing this. I would like to remove per document dependencies as the nodedef for a given node should be discoverable without the original document.

Reply:
Note that I did not remove the use of a "search stage", it's just renamed to stage here. Any nodedefs previously loaded will be found below in the call `stage->findChildByName(nodedefName)`. Since all nodedefs are loaded in before any other element, all existing nodedefs will be available.

However, like you noted we don't support loading a node instance if its nodedef is defined in another document that has not been loaded yet. But that is a use case I don't think we should support. It's the API users responsibility to load in any dependencies first. Otherwise we're back at decoupling node instances from their nodedefs, as in MaterialXCore, which we wanted to move away from.